packages/spacecat-shared-rum-api-client/src/functions/formVitals.js

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+const FORM_SOURCE = ['.form', '.marketo', '.marketo-form'];
+const BOT = 'bot';
+const CHECKPOINT_MAPPING = { formsubmit: 'formsubmit', viewblock: 'formview', click: 'formengagement' };
+
+function initializeResult(url, pageViews) {
+  return {
+    url,
+    formsubmit: {},
+    formview: {},
+    formengagement: {},
+    pageview: pageViews[url],
+  };
+}
+
+function collectFormVitals(bundles, pageViews) {
+  const results = {};
+  for (const bundle of bundles) {
+    const {
+      url, userAgent, weight, events,
+    } = bundle;
+
+    if (userAgent && !userAgent.startsWith(BOT)) {
+      // Track if each condition has been processed for this event
+      const processedCheckpoints = {
+        viewblock: false,
+        formsubmit: false,
+        click: false,
+      };
+
+      // Process each event within the bundle
+      for (const event of events) {
+        const { checkpoint, source } = event;
+
+        // Only process the checkpoint once per event
+        if (!processedCheckpoints[checkpoint]) {
+          if ((checkpoint === 'viewblock' && FORM_SOURCE.includes(source)) || (checkpoint === 'formsubmit') || (checkpoint === 'click' && source && /\bform\b/.test(source.toLowerCase()))) {
```

Comment:
nitpick: we can make it more readable

```
const isFormViewEvent({checkpoint, source}) = 
const isFormClickEvent({checkpoint, source}) =
const isFormSubmitEvent({checkpoint, source}) =
```

Reply:
@05rahulsingh I'm little confused on this logic. Why do you need to process the checkpoints only once? 
If there're two forms on the page, this condition `(checkpoint === 'viewblock' && FORM_SOURCE.includes(source))` can be true twice, right?
IMHO, you should process every event and capture all the possible data

Reply:
And also better to create helper functions with readable names like
`(checkpoint === 'viewblock' && FORM_SOURCE.includes(source))` => `isFormViewed()` etc

Reply:
_just curious:_ Looks like you're only collecting information about the form, wouldn't the interactive elements like `input/select` inside the form, are of interest to you?

Reply:
@rpapani  As in first phase no. As of now we will leverage this invormation in existing cwv audit worker to check for each url whether it has forms embedded. Afterwards we will write our own reporting worker on the same lines of cwx reporting worker to get the reports of url's which has low cwv and has forms embedded in it.

Reply:
> @05rahulsingh I'm little confused on this logic. Why do you need to process the checkpoints only once? If there're two forms on the page, this condition `(checkpoint === 'viewblock' && FORM_SOURCE.includes(source))` can be true twice, right? IMHO, you should process every event and capture all the possible data

viewblock is removed from the check.
formsubmit and click should be modified later based on source and target checks. As of now for this scope its been used to identify whether the page has form or not.
formsubmit can have multiple entries for same form eg submit gives error that's why we are considering single entry, this will be enhanced based on source and target check to get that submit has errors that's why there are multiple entires for form submit with same source and target.




> And also better to create helper functions with readable names like `(checkpoint === 'viewblock' && FORM_SOURCE.includes(source))` => `isFormViewed()` etc

done

Reply:
@05rahulsingh 
> formsubmit can have multiple entries for same form eg submit gives error that's why we are considering single entry,

if multiple formsubmit are there, how you're sure that the entry you're looking at is the successful one, it could be a failure as well right? Or you're only interested in, if the form has been submitted at least once or not, without depending on the result?

Reply:
As of now we are interested in the later one -  if the form has been submitted at least once or not, without depending on the result

Reply:
@rpapani Generally formsubmit is triggered only after validations are passed. Only for 1 customer (out of 10 that we have evaluatated till now) we have noticed that formsubmit are triggered when there are errors as well. We are considering that as an exception. If it becomes a common occurrence with other customers we will revisit.